packages/vite/src/node/config.ts

```diff
@@ -960,47 +965,53 @@ export async function resolveConfig(
   }
 
   // The client and ssr environment configs can't be removed by the user in the config hook
-  if (
-    !config.environments ||
-    !config.environments.client ||
-    (!config.environments.ssr && !isBuild)
-  ) {
+  if (!config.$client || (!config.$ssr && !isBuild)) {
     throw new Error(
       'Required environments configuration were stripped out in the config hook',
     )
   }
 
   // Merge default environment config values
+  const environments: Record<string, EnvironmentOptions> = {}
   const defaultEnvironmentOptions = getDefaultEnvironmentOptions(config)
-  for (const name of Object.keys(config.environments)) {
-    config.environments[name] = mergeConfig(
-      defaultEnvironmentOptions,
-      config.environments[name],
-    )
+  for (const name of Object.keys(config)) {
+    if (name.startsWith('$')) {
+      const environmentName = name as `$${string}`
```

Comment:
Maybe we should also disallow naming environments with just `$`, but I think we can think about this later.